app/services/completions_decider.rb

```diff
@@ -0,0 +1,34 @@
+class CompletionsDecider
+  def initialize(user_agent:, request_url:)
+    @user_agent = user_agent
+    @request_url = request_url
+  end
+
+  def continue_to_sp?
+    user_agent_matches_redirect_uri?
+  end
+
+  private
+
+  attr_reader :user_agent, :request_url
+
+  def user_agent_matches_redirect_uri?
+    desktop_and_web_redirect_uri? || mobile_and_app_redirect_uri?
```

Comment:
We don't have numbers on it, but I think it's a fairly common scenario. I think you only want to block the continue to SP when the redirect uri is not "http*" and the device is desktop.